src/drivers/spi/bcm283x/bcm283x_spi0.c

```diff
@@ -0,0 +1,217 @@
+#include <framework/mod/options.h>
+#include <hal/reg.h>
+#include <util/log.h>
+#include <util/math.h>
+#include <kernel/irq.h>
+#include <kernel/printk.h>
+#include <asm/delay.h>
+
+#include <drivers/gpio/gpio.h>
+#include <drivers/gpio/bcm283x/bcm283x_gpio.h>
+//#include <drivers/mailbox/bcm2835_mailbox_property.h>
+#include <drivers/spi.h>
+
+#include "bcm283x_spi0.h"
+
+#define PBASE		OPTION_GET(NUMBER,base_addr)
+#define AUXBASE		OPTION_GET(NUMBER,aux_addr)
+
+#define SPI_BUS_CLOCK_DIVISOR 	OPTION_GET(NUMBER, spi_bus_clock_divisor)
+
+// Toggle an extra GPIO pin at various points to debug/test timing
+#if 0
+	#define DRIVER_TESTING
+	#define PIN ( 1 << 25 )
+#endif
+
+// Errata to BCM2835 behaviour: documentation states that the SPI0 DLEN register is only used for DMA. 
+// However, even when DMA is not being utilized, setting it from a value != 0 or 1 gets rid of an 
+// excess idle clock cycle that is present when transmitting each byte. (by default in Polled SPI 
+// Mode each 8 bits transfer in 9 clocks) With DLEN=2 each byte is clocked to the bus in 8 cycles, 
+// observed to improve max throughput from 56.8mbps to 63.3mbps (+11.4%, quite close to the 
+// theoretical +12.5%)
+// https://www.raspberrypi.org/forums/viewtopic.php?f=44&t=181154
+#define DLEN_NO_DMA_VALUE 2;
+
+typedef struct {
+	uint32_t cs; /* SPI Master Control and Status */
+	
+/* RW: DMA Mode (DMAEN set)
+	If TA is clear, the first 32-bit write to this register will control 
+	SPIDLEN and SPICS. Subsequent reads and writes will be taken as 
+	four-byte data words to be read/written to the FIFOs Poll/Interrupt 
+	Mode (DMAEN clear, TA set) Writes to the register write bytes to TX FIFO. 
+	Reads from register read bytes from the RX FIFO
+ */
+	uint32_t fifo; /* SPI Master TX and RX FIFOs */
+
+/* RW: 15:0 Clock Divider SCLK = Core Clock / CDIV
+	If CDIV is set to 0, the divisor is 65536. The divisor must be a power of 2.
+	Odd numbers rounded down. The maximum SPI clock rate is of the APB clock.
+ */
+	uint32_t clk; /* SPI Master Clock Divider */
+
+/* RW: 15:0 Data Length - The number of bytes to transfer.
+	This field is only valid for DMA mode (DMAEN set) and controls how many bytes 
+	to transmit (and therefore receive).
+ */	 
+	uint32_t dlen; /* SPI Master Data Length */
+
+/* RW: 3:0 This sets the Output Hold delay in APB clocks. A value of 0 causes a 1 clock delay. */
+	uint32_t ltoh; /* SPI LOSSI mode TOH */
+
+/* RW: This register controls the generation of the DREQ and Panic signals to an 
+	external DMA engine The DREQ signals are generated when the FIFOs reach their 
+	defined levels and need servicing. The Panic signals instruct the external DMA engine 
+	to raise the priority of its AXI requests.*/
+	uint32_t dc; /* SPI DMA DREQ Controls */
+} Bcm283x_spi0;
+
+#define REGS_SPI0 ((Bcm283x_spi0 *)(PBASE))
+#define BCM283X_SPI0_TX_FIFO_LEN (16*sizeof(uint32_t))
+#define BCM283X_SPI0_RX_FIFO_LEN (16*sizeof(uint32_t))
+
+
+/* // Timing, to use elswhere, later
```

Comment:
Is it for the future?

Reply:
Will remove, it will go into another driver